### Module 2
### Session 2 
### Data Cleaning

In [1]:
import pandas as pd

In [2]:
laptops = pd.read_csv('laptops.csv', encoding = 'Windows-1251')

In [3]:
laptops

Manufacturer                           Model Name            Category  \
0           Apple                          MacBook Pro           Ultrabook   
1           Apple                          Macbook Air           Ultrabook   
2              HP                               250 G6            Notebook   
3           Apple                          MacBook Pro           Ultrabook   
4           Apple                          MacBook Pro           Ultrabook   
...           ...                                  ...                 ...   
1298       Lenovo                       Yoga 500-14ISK  2 in 1 Convertible   
1299       Lenovo                       Yoga 900-13ISK  2 in 1 Convertible   
1300       Lenovo                   IdeaPad 100S-14IBR            Notebook   
1301           HP  15-AC110nv (i7-6500U/6GB/1TB/Radeon            Notebook   
1302         Asus  X553SA-XX031T (N3050/4GB/500GB/W10)            Notebook   

     Screen Size                                      Screen  \
0          13.3"          IPS Panel Retina Display 2560x1600   
1          13.3"                                    1440x900   
2          15.6"                           Full HD 1920x1080   
3          15.4"          IPS Panel Retina Display 2880x1800   
4          13.3"          IPS Panel Retina Display 2560x1600   
...          ...                                         ...   
1298       14.0"   IPS Panel Full HD / Touchscreen 1920x1080   
1299       13.3"  IPS Panel Quad HD+ / Touchscreen 3200x1800   
1300       14.0"                                    1366x768   
1301       15.6"                                    1366x768   
1302       15.6"                                    1366x768   

                                       CPU   RAM              Storage  \
0                     Intel Core i5 2.3GHz   8GB            128GB SSD   
1                     Intel Core i5 1.8GHz   8GB  128GB Flash Storage   
2               Intel Core i5 7200U 2.5GHz   8GB            256GB SSD   
3                     Intel Core i7 2.7GHz  16GB            512GB SSD   
4                     Intel Core i5 3.1GHz   8GB            256GB SSD   
...                                    ...   ...                  ...   
1298            Intel Core i7 6500U 2.5GHz   4GB            128GB SSD   
1299            Intel Core i7 6500U 2.5GHz  16GB            512GB SSD   
1300  Intel Celeron Dual Core N3050 1.6GHz   2GB   64GB Flash Storage   
1301            Intel Core i7 6500U 2.5GHz   6GB              1TB HDD   
1302  Intel Celeron Dual Core N3050 1.6GHz   4GB            500GB HDD   

                               GPU Operating System Operating System Version  \
0     Intel Iris Plus Graphics 640            macOS                      NaN   
1           Intel HD Graphics 6000            macOS                      NaN   
2            Intel HD Graphics 620            No OS                      NaN   
3               AMD Radeon Pro 455            macOS                      NaN   
4     Intel Iris Plus Graphics 650            macOS                      NaN   
...                            ...              ...                      ...   
1298         Intel HD Graphics 520          Windows                       10   
1299         Intel HD Graphics 520          Windows                       10   
1300             Intel HD Graphics          Windows                       10   
1301            AMD Radeon R5 M330          Windows                       10   
1302             Intel HD Graphics          Windows                       10   

      Weight Price (Euros)  
0     1.37kg       1339,69  
1     1.34kg        898,94  
2     1.86kg        575,00  
3     1.83kg       2537,45  
4     1.37kg       1803,60  
...      ...           ...  
1298   1.8kg        638,00  
1299   1.3kg       1499,00  
1300   1.5kg        229,00  
1301  2.19kg        764,00  
1302   2.2kg        369,00  

[1303 rows x 13 columns]

In [4]:
laptops.columns

Index(['Manufacturer', 'Model Name', 'Category', 'Screen Size', 'Screen',
       'CPU', 'RAM', ' Storage', 'GPU', 'Operating System',
       'Operating System Version', 'Weight', 'Price (Euros)'],
      dtype='object')

In [5]:
laptops.columns = laptops.columns.str.strip().str.replace('(', '').str.replace(')','').str.lower().str.replace(' ','_')\
.str.replace('operating_system','os')       #Replacing Caps with lower case and spaces with _

In [6]:
laptops.columns

Index(['manufacturer', 'model_name', 'category', 'screen_size', 'screen',
       'cpu', 'ram', 'storage', 'gpu', 'os', 'os_version', 'weight',
       'price_euros'],
      dtype='object')

In [7]:
laptops["screen_size"] = laptops["screen_size"].str.strip('"').astype(float) 

#eliminating " sign from inches and typecasting

In [8]:
laptops["screen_size"]

0       13.3
1       13.3
2       15.6
3       15.4
4       13.3
        ... 
1298    14.0
1299    13.3
1300    14.0
1301    15.6
1302    15.6
Name: screen_size, Length: 1303, dtype: float64

In [9]:
#laptops = laptops.rename({"screen_size":"screen_size_inches"}, axis=1)

laptops.rename({"screen_size":"screen_size_inches"}, axis=1, inplace= True)   #inplace updates the environment where its working

In [10]:
laptops

manufacturer                           model_name            category  \
0           Apple                          MacBook Pro           Ultrabook   
1           Apple                          Macbook Air           Ultrabook   
2              HP                               250 G6            Notebook   
3           Apple                          MacBook Pro           Ultrabook   
4           Apple                          MacBook Pro           Ultrabook   
...           ...                                  ...                 ...   
1298       Lenovo                       Yoga 500-14ISK  2 in 1 Convertible   
1299       Lenovo                       Yoga 900-13ISK  2 in 1 Convertible   
1300       Lenovo                   IdeaPad 100S-14IBR            Notebook   
1301           HP  15-AC110nv (i7-6500U/6GB/1TB/Radeon            Notebook   
1302         Asus  X553SA-XX031T (N3050/4GB/500GB/W10)            Notebook   

      screen_size_inches                                      screen  \
0                   13.3          IPS Panel Retina Display 2560x1600   
1                   13.3                                    1440x900   
2                   15.6                           Full HD 1920x1080   
3                   15.4          IPS Panel Retina Display 2880x1800   
4                   13.3          IPS Panel Retina Display 2560x1600   
...                  ...                                         ...   
1298                14.0   IPS Panel Full HD / Touchscreen 1920x1080   
1299                13.3  IPS Panel Quad HD+ / Touchscreen 3200x1800   
1300                14.0                                    1366x768   
1301                15.6                                    1366x768   
1302                15.6                                    1366x768   

                                       cpu   ram              storage  \
0                     Intel Core i5 2.3GHz   8GB            128GB SSD   
1                     Intel Core i5 1.8GHz   8GB  128GB Flash Storage   
2               Intel Core i5 7200U 2.5GHz   8GB            256GB SSD   
3                     Intel Core i7 2.7GHz  16GB            512GB SSD   
4                     Intel Core i5 3.1GHz   8GB            256GB SSD   
...                                    ...   ...                  ...   
1298            Intel Core i7 6500U 2.5GHz   4GB            128GB SSD   
1299            Intel Core i7 6500U 2.5GHz  16GB            512GB SSD   
1300  Intel Celeron Dual Core N3050 1.6GHz   2GB   64GB Flash Storage   
1301            Intel Core i7 6500U 2.5GHz   6GB              1TB HDD   
1302  Intel Celeron Dual Core N3050 1.6GHz   4GB            500GB HDD   

                               gpu       os os_version  weight price_euros  
0     Intel Iris Plus Graphics 640    macOS        NaN  1.37kg     1339,69  
1           Intel HD Graphics 6000    macOS        NaN  1.34kg      898,94  
2            Intel HD Graphics 620    No OS        NaN  1.86kg      575,00  
3               AMD Radeon Pro 455    macOS        NaN  1.83kg     2537,45  
4     Intel Iris Plus Graphics 650    macOS        NaN  1.37kg     1803,60  
...                            ...      ...        ...     ...         ...  
1298         Intel HD Graphics 520  Windows         10   1.8kg      638,00  
1299         Intel HD Graphics 520  Windows         10   1.3kg     1499,00  
1300             Intel HD Graphics  Windows         10   1.5kg      229,00  
1301            AMD Radeon R5 M330  Windows         10  2.19kg      764,00  
1302             Intel HD Graphics  Windows         10   2.2kg      369,00  

[1303 rows x 13 columns]

In [11]:
laptops["screen_size_inches"].mean()

15.017191097467379

In [12]:
#Avg Screen size is 15inches

In [13]:
laptops["weight"] = laptops["weight"].str.strip("kg").str.strip("kgs").astype(float)

In [14]:
laptops.rename({"weight":"weight_kg"}, axis=1, inplace= True)

In [15]:
laptops["weight_kg"]

0       1.37
1       1.34
2       1.86
3       1.83
4       1.37
        ... 
1298    1.80
1299    1.30
1300    1.50
1301    2.19
1302    2.20
Name: weight_kg, Length: 1303, dtype: float64

In [16]:
laptops

manufacturer                           model_name            category  \
0           Apple                          MacBook Pro           Ultrabook   
1           Apple                          Macbook Air           Ultrabook   
2              HP                               250 G6            Notebook   
3           Apple                          MacBook Pro           Ultrabook   
4           Apple                          MacBook Pro           Ultrabook   
...           ...                                  ...                 ...   
1298       Lenovo                       Yoga 500-14ISK  2 in 1 Convertible   
1299       Lenovo                       Yoga 900-13ISK  2 in 1 Convertible   
1300       Lenovo                   IdeaPad 100S-14IBR            Notebook   
1301           HP  15-AC110nv (i7-6500U/6GB/1TB/Radeon            Notebook   
1302         Asus  X553SA-XX031T (N3050/4GB/500GB/W10)            Notebook   

      screen_size_inches                                      screen  \
0                   13.3          IPS Panel Retina Display 2560x1600   
1                   13.3                                    1440x900   
2                   15.6                           Full HD 1920x1080   
3                   15.4          IPS Panel Retina Display 2880x1800   
4                   13.3          IPS Panel Retina Display 2560x1600   
...                  ...                                         ...   
1298                14.0   IPS Panel Full HD / Touchscreen 1920x1080   
1299                13.3  IPS Panel Quad HD+ / Touchscreen 3200x1800   
1300                14.0                                    1366x768   
1301                15.6                                    1366x768   
1302                15.6                                    1366x768   

                                       cpu   ram              storage  \
0                     Intel Core i5 2.3GHz   8GB            128GB SSD   
1                     Intel Core i5 1.8GHz   8GB  128GB Flash Storage   
2               Intel Core i5 7200U 2.5GHz   8GB            256GB SSD   
3                     Intel Core i7 2.7GHz  16GB            512GB SSD   
4                     Intel Core i5 3.1GHz   8GB            256GB SSD   
...                                    ...   ...                  ...   
1298            Intel Core i7 6500U 2.5GHz   4GB            128GB SSD   
1299            Intel Core i7 6500U 2.5GHz  16GB            512GB SSD   
1300  Intel Celeron Dual Core N3050 1.6GHz   2GB   64GB Flash Storage   
1301            Intel Core i7 6500U 2.5GHz   6GB              1TB HDD   
1302  Intel Celeron Dual Core N3050 1.6GHz   4GB            500GB HDD   

                               gpu       os os_version  weight_kg price_euros  
0     Intel Iris Plus Graphics 640    macOS        NaN       1.37     1339,69  
1           Intel HD Graphics 6000    macOS        NaN       1.34      898,94  
2            Intel HD Graphics 620    No OS        NaN       1.86      575,00  
3               AMD Radeon Pro 455    macOS        NaN       1.83     2537,45  
4     Intel Iris Plus Graphics 650    macOS        NaN       1.37     1803,60  
...                            ...      ...        ...        ...         ...  
1298         Intel HD Graphics 520  Windows         10       1.80      638,00  
1299         Intel HD Graphics 520  Windows         10       1.30     1499,00  
1300             Intel HD Graphics  Windows         10       1.50      229,00  
1301            AMD Radeon R5 M330  Windows         10       2.19      764,00  
1302             Intel HD Graphics  Windows         10       2.20      369,00  

[1303 rows x 13 columns]

In [17]:
laptops["weight_kg"].mean()

2.0387336914811973

In [18]:
#Avg weight is around 2kgs

In [19]:
laptops["price_euros"] = laptops["price_euros"].str.replace(",",".").astype(float)

In [20]:
laptops["price_euros"]

0       1339.69
1        898.94
2        575.00
3       2537.45
4       1803.60
         ...   
1298     638.00
1299    1499.00
1300     229.00
1301     764.00
1302     369.00
Name: price_euros, Length: 1303, dtype: float64

In [21]:
laptops["price_euros"].mean()

1123.6869915579432

In [22]:
#Avg price is around 1125 euros

In [23]:
laptops["ram"] = laptops["ram"].str.strip("GB").astype(float)

In [24]:
laptops["ram"]

0        8.0
1        8.0
2        8.0
3       16.0
4        8.0
        ... 
1298     4.0
1299    16.0
1300     2.0
1301     6.0
1302     4.0
Name: ram, Length: 1303, dtype: float64

In [25]:
laptops["ram"].mean()

8.382194934765925

In [26]:
#Avg RAM size is 8GB

In [27]:
gpu_split = laptops["gpu"].str.split(n=1, expand=True)   #splits from 1st space    #splits in 2 different columns

In [28]:
laptops["gpu_company"] = gpu_split[0]
laptops["gpu_model"] = gpu_split[1]

In [29]:
laptops

manufacturer                           model_name            category  \
0           Apple                          MacBook Pro           Ultrabook   
1           Apple                          Macbook Air           Ultrabook   
2              HP                               250 G6            Notebook   
3           Apple                          MacBook Pro           Ultrabook   
4           Apple                          MacBook Pro           Ultrabook   
...           ...                                  ...                 ...   
1298       Lenovo                       Yoga 500-14ISK  2 in 1 Convertible   
1299       Lenovo                       Yoga 900-13ISK  2 in 1 Convertible   
1300       Lenovo                   IdeaPad 100S-14IBR            Notebook   
1301           HP  15-AC110nv (i7-6500U/6GB/1TB/Radeon            Notebook   
1302         Asus  X553SA-XX031T (N3050/4GB/500GB/W10)            Notebook   

      screen_size_inches                                      screen  \
0                   13.3          IPS Panel Retina Display 2560x1600   
1                   13.3                                    1440x900   
2                   15.6                           Full HD 1920x1080   
3                   15.4          IPS Panel Retina Display 2880x1800   
4                   13.3          IPS Panel Retina Display 2560x1600   
...                  ...                                         ...   
1298                14.0   IPS Panel Full HD / Touchscreen 1920x1080   
1299                13.3  IPS Panel Quad HD+ / Touchscreen 3200x1800   
1300                14.0                                    1366x768   
1301                15.6                                    1366x768   
1302                15.6                                    1366x768   

                                       cpu   ram              storage  \
0                     Intel Core i5 2.3GHz   8.0            128GB SSD   
1                     Intel Core i5 1.8GHz   8.0  128GB Flash Storage   
2               Intel Core i5 7200U 2.5GHz   8.0            256GB SSD   
3                     Intel Core i7 2.7GHz  16.0            512GB SSD   
4                     Intel Core i5 3.1GHz   8.0            256GB SSD   
...                                    ...   ...                  ...   
1298            Intel Core i7 6500U 2.5GHz   4.0            128GB SSD   
1299            Intel Core i7 6500U 2.5GHz  16.0            512GB SSD   
1300  Intel Celeron Dual Core N3050 1.6GHz   2.0   64GB Flash Storage   
1301            Intel Core i7 6500U 2.5GHz   6.0              1TB HDD   
1302  Intel Celeron Dual Core N3050 1.6GHz   4.0            500GB HDD   

                               gpu       os os_version  weight_kg  \
0     Intel Iris Plus Graphics 640    macOS        NaN       1.37   
1           Intel HD Graphics 6000    macOS        NaN       1.34   
2            Intel HD Graphics 620    No OS        NaN       1.86   
3               AMD Radeon Pro 455    macOS        NaN       1.83   
4     Intel Iris Plus Graphics 650    macOS        NaN       1.37   
...                            ...      ...        ...        ...   
1298         Intel HD Graphics 520  Windows         10       1.80   
1299         Intel HD Graphics 520  Windows         10       1.30   
1300             Intel HD Graphics  Windows         10       1.50   
1301            AMD Radeon R5 M330  Windows         10       2.19   
1302             Intel HD Graphics  Windows         10       2.20   

      price_euros gpu_company               gpu_model  
0         1339.69       Intel  Iris Plus Graphics 640  
1          898.94       Intel        HD Graphics 6000  
2          575.00       Intel         HD Graphics 620  
3         2537.45         AMD          Radeon Pro 455  
4         1803.60       Intel  Iris Plus Graphics 650  
...           ...         ...                     ...  
1298       638.00       Intel         HD Graphics 520  
1299      1499.00       Intel         HD Graphics

In [30]:
laptops["gpu_company"].value_counts()

gpu_company
Intel     722
Nvidia    400
AMD       180
ARM         1
Name: count, dtype: int64

In [31]:
laptops["gpu_company"].value_counts(normalize=True)*100

gpu_company
Intel     55.410591
Nvidia    30.698388
AMD       13.814275
ARM        0.076746
Name: proportion, dtype: float64

In [32]:
laptops_gpu = laptops[(laptops["gpu_company"] != "Intel") & (laptops["gpu_company"] != "ARM")]

In [33]:
laptops_gpu["gpu_company"].value_counts(normalize=True)*100

gpu_company
Nvidia    68.965517
AMD       31.034483
Name: proportion, dtype: float64

In [34]:
laptops_gpu["gpu_model"].value_counts()

gpu_model
GeForce GTX 1050    66
GeForce GTX 1060    48
GeForce 940MX       43
Radeon 530          41
GeForce GTX 1070    30
                    ..
Quadro M620M         1
Radeon R5 520        1
Radeon R7            1
Radeon 540           1
GeForce 960M         1
Name: count, Length: 86, dtype: int64

In [35]:
cpu_split = laptops["cpu"].str.split(n=1, expand= True)

In [36]:
laptops["cpu_company"] = cpu_split[0]
laptops["cpu_model"] = cpu_split[1]

In [37]:
laptops

manufacturer                           model_name            category  \
0           Apple                          MacBook Pro           Ultrabook   
1           Apple                          Macbook Air           Ultrabook   
2              HP                               250 G6            Notebook   
3           Apple                          MacBook Pro           Ultrabook   
4           Apple                          MacBook Pro           Ultrabook   
...           ...                                  ...                 ...   
1298       Lenovo                       Yoga 500-14ISK  2 in 1 Convertible   
1299       Lenovo                       Yoga 900-13ISK  2 in 1 Convertible   
1300       Lenovo                   IdeaPad 100S-14IBR            Notebook   
1301           HP  15-AC110nv (i7-6500U/6GB/1TB/Radeon            Notebook   
1302         Asus  X553SA-XX031T (N3050/4GB/500GB/W10)            Notebook   

      screen_size_inches                                      screen  \
0                   13.3          IPS Panel Retina Display 2560x1600   
1                   13.3                                    1440x900   
2                   15.6                           Full HD 1920x1080   
3                   15.4          IPS Panel Retina Display 2880x1800   
4                   13.3          IPS Panel Retina Display 2560x1600   
...                  ...                                         ...   
1298                14.0   IPS Panel Full HD / Touchscreen 1920x1080   
1299                13.3  IPS Panel Quad HD+ / Touchscreen 3200x1800   
1300                14.0                                    1366x768   
1301                15.6                                    1366x768   
1302                15.6                                    1366x768   

                                       cpu   ram              storage  \
0                     Intel Core i5 2.3GHz   8.0            128GB SSD   
1                     Intel Core i5 1.8GHz   8.0  128GB Flash Storage   
2               Intel Core i5 7200U 2.5GHz   8.0            256GB SSD   
3                     Intel Core i7 2.7GHz  16.0            512GB SSD   
4                     Intel Core i5 3.1GHz   8.0            256GB SSD   
...                                    ...   ...                  ...   
1298            Intel Core i7 6500U 2.5GHz   4.0            128GB SSD   
1299            Intel Core i7 6500U 2.5GHz  16.0            512GB SSD   
1300  Intel Celeron Dual Core N3050 1.6GHz   2.0   64GB Flash Storage   
1301            Intel Core i7 6500U 2.5GHz   6.0              1TB HDD   
1302  Intel Celeron Dual Core N3050 1.6GHz   4.0            500GB HDD   

                               gpu       os os_version  weight_kg  \
0     Intel Iris Plus Graphics 640    macOS        NaN       1.37   
1           Intel HD Graphics 6000    macOS        NaN       1.34   
2            Intel HD Graphics 620    No OS        NaN       1.86   
3               AMD Radeon Pro 455    macOS        NaN       1.83   
4     Intel Iris Plus Graphics 650    macOS        NaN       1.37   
...                            ...      ...        ...        ...   
1298         Intel HD Graphics 520  Windows         10       1.80   
1299         Intel HD Graphics 520  Windows         10       1.30   
1300             Intel HD Graphics  Windows         10       1.50   
1301            AMD Radeon R5 M330  Windows         10       2.19   
1302             Intel HD Graphics  Windows         10       2.20   

      price_euros gpu_company               gpu_model cpu_company  \
0         1339.69       Intel  Iris Plus Graphics 640       Intel   
1          898.94       Intel        HD Graphics 6000       Intel   
2          575.00       Intel         HD Graphics 620       Intel   
3         2537.45         AMD          Radeon Pro 455       Intel   
4         1803.60       Intel  Iris Plus Graphics 650       Intel   
...           ...         ...                     ...         ...   
1298       638

In [38]:
laptops["cpu_company"].value_counts(normalize= True)*100

cpu_company
Intel      95.165004
AMD         4.758250
Samsung     0.076746
Name: proportion, dtype: float64

In [39]:
laptops["cpu_model"].value_counts()

cpu_model
Core i5 7200U 2.5GHz       190
Core i7 7700HQ 2.8GHz      146
Core i7 7500U 2.7GHz       134
Core i7 8550U 1.8GHz        73
Core i5 8250U 1.6GHz        72
                          ... 
Core M M3-6Y30 0.9GHz        1
A9-Series 9420 2.9GHz        1
Core i3 6006U 2.2GHz         1
A6-Series 7310 2GHz          1
Xeon E3-1535M v6 3.1GHz      1
Name: count, Length: 118, dtype: int64

### Hence its prefereable to include following specs
1) Nvidia company gpu
2) GeForce GTX 1050 or GeForce GTX 1060 gpu model
3) 8GB RAM
4) Intel processor with speed > 2GHz
5) ~15inches screen size
6) 2kgs weight
7) Price of 1125 Euros

In [40]:
laptops["os"].unique()

array(['macOS', 'No OS', 'Windows', 'Mac OS', 'Linux', 'Android',
       'Chrome OS'], dtype=object)

In [41]:
#laptops["os"].str.replace("Mac OS", "mac OS")

laptops.loc[laptops["os"] == "Mac OS", "os"] 

6       Mac OS
26      Mac OS
34      Mac OS
794     Mac OS
1069    Mac OS
1193    Mac OS
1210    Mac OS
1234    Mac OS
Name: os, dtype: object

In [42]:
laptops.loc[laptops["os"] == "Mac OS", "os"] = "macOS"     #Alternate way of replacing by first accessing and then using loc

In [43]:
laptops["os"].unique()

array(['macOS', 'No OS', 'Windows', 'Linux', 'Android', 'Chrome OS'],
      dtype=object)

In [44]:
laptops["os"].value_counts()

os
Windows      1125
No OS          66
Linux          62
Chrome OS      27
macOS          21
Android         2
Name: count, dtype: int64

### Missing values

In [45]:
laptops.isnull().sum()

manufacturer            0
model_name              0
category                0
screen_size_inches      0
screen                  0
cpu                     0
ram                     0
storage                 0
gpu                     0
os                      0
os_version            170
weight_kg               0
price_euros             0
gpu_company             0
gpu_model               0
cpu_company             0
cpu_model               0
dtype: int64

In [46]:
laptops.loc[laptops["os_version"].isnull()]

manufacturer      model_name   category  screen_size_inches  \
0           Apple     MacBook Pro  Ultrabook                13.3   
1           Apple     Macbook Air  Ultrabook                13.3   
2              HP          250 G6   Notebook                15.6   
3           Apple     MacBook Pro  Ultrabook                15.4   
4           Apple     MacBook Pro  Ultrabook                13.3   
...           ...             ...        ...                 ...   
1267         Dell   Inspiron 3567   Notebook                15.6   
1277         Acer  Aspire ES1-531   Notebook                15.6   
1281         Dell   Inspiron 3567   Notebook                15.6   
1291         Acer  Aspire ES1-531   Notebook                15.6   
1295         Dell   Inspiron 3567   Notebook                15.6   

                                  screen  \
0     IPS Panel Retina Display 2560x1600   
1                               1440x900   
2                      Full HD 1920x1080   
3     IPS Panel Retina Display 2880x1800   
4     IPS Panel Retina Display 2560x1600   
...                                  ...   
1267                            1366x768   
1277                            1366x768   
1281                            1366x768   
1291                            1366x768   
1295                            1366x768   

                                       cpu   ram              storage  \
0                     Intel Core i5 2.3GHz   8.0            128GB SSD   
1                     Intel Core i5 1.8GHz   8.0  128GB Flash Storage   
2               Intel Core i5 7200U 2.5GHz   8.0            256GB SSD   
3                     Intel Core i7 2.7GHz  16.0            512GB SSD   
4                     Intel Core i5 3.1GHz   8.0            256GB SSD   
...                                    ...   ...                  ...   
1267            Intel Core i7 7500U 2.7GHz   8.0              1TB HDD   
1277  Intel Celeron Dual Core N3060 1.6GHz   4.0            500GB HDD   
1281            Intel Core i7 7500U 2.7GHz   8.0              1TB HDD   
1291  Intel Celeron Dual Core N3060 1.6GHz   4.0            500GB HDD   
1295            Intel Core i7 7500U 2.7GHz   8.0              1TB HDD   

                               gpu     os os_version  weight_kg  price_euros  \
0     Intel Iris Plus Graphics 640  macOS        NaN       1.37      1339.69   
1           Intel HD Graphics 6000  macOS        NaN       1.34       898.94   
2            Intel HD Graphics 620  No OS        NaN       1.86       575.00   
3               AMD Radeon Pro 455  macOS        NaN       1.83      2537.45   
4     Intel Iris Plus Graphics 650  macOS        NaN       1.37      1803.60   
...                            ...    ...        ...        ...          ...   
1267            AMD Radeon R5 M430  Linux        NaN       2.30       805.99   
1277         Intel HD Graphics 400  Linux        NaN       2.40       289.00   
1281            AMD Radeon R5 M430  Linux        NaN       2.30       805.99   
1291         Intel HD Graphics 400  Linux        NaN       2.40       289.00   
1295            AMD Radeon R5 M430  Linux        NaN       2.30       805.99   

     gpu_company               gpu_model cpu_company  \
0          Intel  Iris Plus Graphics 640       Intel   
1          Intel        HD Graphics 6000       Intel   
2          Intel         HD Graphics 620       Intel   
3            AMD          Radeon Pro 455       Intel   
4          Intel  Iris Plus Graphics 650       Intel   
...          ...                     ...         ...   
1267         AMD          Radeon R5 M430       Intel   
1277       Intel         HD Graphics 400       Intel   
1281         AMD          Radeon R5 M430       Intel   
1291       Intel         HD Graphics 400       Intel   
1295         AMD          Radeon R5 M430       Intel   

                           cpu_model  
0                     Core i5 2.3GHz  
1                     Core i5 1.8GHz  
2               Core i5 7200U 2

In [47]:
laptops.loc[laptops["os_version"].isnull(), "os"].value_counts()

os
No OS        66
Linux        62
Chrome OS    27
macOS        13
Android       2
Name: count, dtype: int64

In [48]:
laptops.loc[laptops["os_version"].isnull(), "os_version"] = "not specified"

In [49]:
laptops.isnull().sum()

manufacturer          0
model_name            0
category              0
screen_size_inches    0
screen                0
cpu                   0
ram                   0
storage               0
gpu                   0
os                    0
os_version            0
weight_kg             0
price_euros           0
gpu_company           0
gpu_model             0
cpu_company           0
cpu_model             0
dtype: int64

In [50]:
laptops

manufacturer                           model_name            category  \
0           Apple                          MacBook Pro           Ultrabook   
1           Apple                          Macbook Air           Ultrabook   
2              HP                               250 G6            Notebook   
3           Apple                          MacBook Pro           Ultrabook   
4           Apple                          MacBook Pro           Ultrabook   
...           ...                                  ...                 ...   
1298       Lenovo                       Yoga 500-14ISK  2 in 1 Convertible   
1299       Lenovo                       Yoga 900-13ISK  2 in 1 Convertible   
1300       Lenovo                   IdeaPad 100S-14IBR            Notebook   
1301           HP  15-AC110nv (i7-6500U/6GB/1TB/Radeon            Notebook   
1302         Asus  X553SA-XX031T (N3050/4GB/500GB/W10)            Notebook   

      screen_size_inches                                      screen  \
0                   13.3          IPS Panel Retina Display 2560x1600   
1                   13.3                                    1440x900   
2                   15.6                           Full HD 1920x1080   
3                   15.4          IPS Panel Retina Display 2880x1800   
4                   13.3          IPS Panel Retina Display 2560x1600   
...                  ...                                         ...   
1298                14.0   IPS Panel Full HD / Touchscreen 1920x1080   
1299                13.3  IPS Panel Quad HD+ / Touchscreen 3200x1800   
1300                14.0                                    1366x768   
1301                15.6                                    1366x768   
1302                15.6                                    1366x768   

                                       cpu   ram              storage  \
0                     Intel Core i5 2.3GHz   8.0            128GB SSD   
1                     Intel Core i5 1.8GHz   8.0  128GB Flash Storage   
2               Intel Core i5 7200U 2.5GHz   8.0            256GB SSD   
3                     Intel Core i7 2.7GHz  16.0            512GB SSD   
4                     Intel Core i5 3.1GHz   8.0            256GB SSD   
...                                    ...   ...                  ...   
1298            Intel Core i7 6500U 2.5GHz   4.0            128GB SSD   
1299            Intel Core i7 6500U 2.5GHz  16.0            512GB SSD   
1300  Intel Celeron Dual Core N3050 1.6GHz   2.0   64GB Flash Storage   
1301            Intel Core i7 6500U 2.5GHz   6.0              1TB HDD   
1302  Intel Celeron Dual Core N3050 1.6GHz   4.0            500GB HDD   

                               gpu       os     os_version  weight_kg  \
0     Intel Iris Plus Graphics 640    macOS  not specified       1.37   
1           Intel HD Graphics 6000    macOS  not specified       1.34   
2            Intel HD Graphics 620    No OS  not specified       1.86   
3               AMD Radeon Pro 455    macOS  not specified       1.83   
4     Intel Iris Plus Graphics 650    macOS  not specified       1.37   
...                            ...      ...            ...        ...   
1298         Intel HD Graphics 520  Windows             10       1.80   
1299         Intel HD Graphics 520  Windows             10       1.30   
1300             Intel HD Graphics  Windows             10       1.50   
1301            AMD Radeon R5 M330  Windows             10       2.19   
1302             Intel HD Graphics  Windows             10       2.20   

      price_euros gpu_company               gpu_model cpu_company  \
0         1339.69       Intel  Iris Plus Graphics 640       Intel   
1          898.94       Intel        HD Graphics 6000       Intel   
2          575.00       Intel         HD Graphics 620       Intel   
3         2537.45         AMD          Radeon Pro 455       Intel   
4         1803.60       Intel  Iris Plus Graphics 650       Intel   
...           ...         ...      